src/main/java/servlets/OAuthCallbackServlet.java

```diff
@@ -0,0 +1,85 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package servlets;
+
+import com.google.gson.JsonParser;
+import java.io.IOException;
+import java.io.FileInputStream;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.net.http.HttpRequest.BodyPublishers;
+import java.net.http.HttpResponse.BodyHandlers;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonElement;
+
+/** A servlet to request access and refresh tokens. */
+@WebServlet("/api/oauth/callback/sheets")
+public class OAuthCallbackServlet extends HttpServlet {
+  // Token request parameters.
+  private final String CLIENT_ID = "client_id=150737768611-svndjtlklolq53g4ass4r3sqal2i31p5.apps.googleusercontent.com";
+  private final String GRANT_TYPE = "grant_type=authorization_code";
+  private final String CODE = "https://accounts.google.com/o/oauth2/v2/auth";
+  private final String TOKEN_URI = "https://oauth2.googleapis.com/token";
+
+  // Figure out a way to secure method to retrieve the client secret.
+  private final String CLIENT_SECRET = "";
+  private final String REDIRECT_URI = "redirect_uri= SOME CALLBACK URI";
+
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String error = request.getParameter("error");
+    String code = request.getParameter("code");
+
+    // Print any error the OAuth provider gave us.
+    if (error != null && !error.isEmpty()) {
+      response.getWriter().print(error);
+      response.setStatus(401);
+      return;
+    }
+
+    // OAuth provider should have given us a code, otherwise something is wrong here
+    if (code == null || code.isEmpty()) {
+      response.setStatus(400);
+      return;
+    }
+
+    String tokenRequestBody = String.format("%s?%s&%s&%s&%s", GRANT_TYPE, CODE,
+      REDIRECT_URI, CLIENT_ID, CLIENT_SECRET);
+
+    // Request the tokens.
+    HttpClient httpClient = HttpClient.newHttpClient();
+    HttpRequest tokenRequest = HttpRequest.newBuilder(URI.create(TOKEN_URI))
+      .header("Content-Type", "application/x-www-form-urlencoded")
+      .POST(BodyPublishers.ofString(tokenRequestBody)).build();
+
+    HttpResponse tokenResponse = httpClient.sendAsync(tokenRequest, BodyHandlers.ofString()).join();
+    String tokenResponseBody = tokenResponse.body().toString();
+
+    // Parses to find access token.
+    JsonObject tokenResponseObj = JsonParser.parseString(tokenResponseBody).getAsJsonObject();
+    JsonElement accessToken = tokenResponseObj.get("access_token");
```

Comment:
make a constant? also for contentType?

Reply:
For my OAuthCallbackServlet PR, I will create a constants class for final variables that are used in both the Login and OAuthCallback Servlets.